pwnlib/heap/glmalloc/malloc_state/bins/parser.py

```diff
@@ -0,0 +1,200 @@
+from pwnlib.heap.glmalloc.bins import SmallBinEntry, LargeBinEntry, UnsortedBinEntry
+from .bins_indexes import \
+    LARGE_BINS_START_INDEX, SMALL_BINS_START_INDEX
+from .bins import Bins
+
+
+class BinsParser:
+    """Class with the logic to parse the `bins` attribute of the
+    `malloc_state` struct.
+
+    Args:
+         pointer_size (int): The pointer size in bytes of the process.
+    """
+
+    def __init__(self, pointer_size):
+        self._pointer_size = pointer_size
+        if pointer_size == 8:
+            self._largebin_index_to_min_size = largebin64_index_to_min_size
+        else:
+            self._largebin_index_to_min_size = largebin32_index_to_min_size
+
+    def parse_from_collection(self, base_address, collection_array):
+        """Returns a FastBinsY object by parsing a `bins` binary array.
+
+        Args:
+            collection_array (bytes): Binary `bins` array of `malloc_state` struct.
+
+        Returns:
+            Bins
+        """
+
+        num_entries = len(collection_array)
+        entries = []
+
+        for i in range(0, num_entries, 2):
+            fd = collection_array[i]
+            bk = collection_array[i + 1]
+            address = base_address + i * self._pointer_size
+
+            index = int(i / 2)
+            entry = self._to_bin_entry(index, address, fd, bk)
+            entries.append(entry)
+
+        return Bins(entries)
+
+    def _to_bin_entry(self, index, address, fd, bk):
+        if index < SMALL_BINS_START_INDEX:
+            return UnsortedBinEntry(address, fd, bk)
+        elif index < LARGE_BINS_START_INDEX:
+            chunks_size = self._bin_index_to_size(index)
+            return SmallBinEntry(address, fd, bk, chunks_size)
+        else:
+            min_chunks_size = self._bin_index_to_size(index)
+            return LargeBinEntry(address, fd, bk, min_chunks_size)
+
+    def _bin_index_to_size(self, index):
+        if index < SMALL_BINS_START_INDEX:
+            return 0
+        elif index < LARGE_BINS_START_INDEX:
+            return (self._pointer_size * 4) + (index - 1) * (self._pointer_size * 2)
+        elif index < 126:
```

Comment:
What is this magic number 126? Needs docs.

Reply:
Done in aedf3b8